hal/src/thumbv6m/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1353 @@
+    }
+}
+
+impl<C, D> Uart<C, D>
+where
+    C: ValidConfig,
+    D: Simplex,
+{
+    /// Disable the UART peripheral and return the underlying [`Config`]
+    #[inline]
+    pub fn disable(self) -> C {
+        let mut config = self.config;
+        config.as_mut()._disable();
+        config
+    }
+
+    /// Update the UART [`Config`]uration.
+    ///
+    /// Calling this method will temporarily disable the SERCOM peripheral, as
+    /// some registers are enable-protected. This may interrupt any ongoing
+    /// transactions.
+    ///
+    /// ```
+    /// use atsamd_hal::sercom::v2::uart::{BaudMode, Oversampling, Uart};
+    /// uart.reconfigure(|c| c.baud(9600, BaudMode::Fractional(Oversampling::Bits16)));
+    /// ```
+    #[inline]
+    pub fn reconfigure<U>(&mut self, update: U)
+    where
+        U: FnOnce(Reconfig<C>) -> Reconfig<C>,
+    {
+        self._reconfigure(update);
+    }
+}
+
+impl<C> Uart<C, Duplex>
+where
+    C: ValidConfig,
+{
+    /// Split the [`Uart`] into [`RxDuplex`] and [`TxDuplex`] halves
+    #[inline]
+    pub fn split(self) -> (Uart<C, RxDuplex>, Uart<C, TxDuplex>) {
+        let config = unsafe { core::ptr::read(&self.config) };
+        (
+            Uart {
+                config: self.config,
+                capability: PhantomData,
+            },
+            Uart {
+                config,
+                capability: PhantomData,
+            },
+        )
+    }
+
+    /// Disable the UART peripheral and return the underlying [`Config`]
+    #[inline]
+    pub fn disable(self) -> C {
+        let mut config = self.config;
+        config.as_mut()._disable();
+        config
+    }
+
+    /// Update the UART [`Config`]uration.
+    ///
+    /// Calling this method will temporarily disable the SERCOM peripheral, as
+    /// some registers are enable-protected. This may interrupt any ongoing
+    /// transactions.
+    ///
+    /// ```
+    /// use atsamd_hal::sercom::v2::uart::{BaudMode, Oversampling, Uart};
+    /// uart.reconfigure(|c| c.baud(9600, BaudMode::Fractional(Oversampling::Bits16)));
+    /// ```
+    #[inline]
+    pub fn reconfigure<F>(&mut self, update: F)
+    where
+        F: FnOnce(Reconfig<C>) -> Reconfig<C>,
+    {
+        self._reconfigure(update);
+    }
+
+    /// Join [`RxDuplex`] and [`TxDuplex`] halves back into a full `Uart<C,
+    /// Duplex>`
+    pub fn join(rx: Uart<C, RxDuplex>, _tx: Uart<C, TxDuplex>) -> Self {
+        Self {
+            config: rx.config,
+            capability: PhantomData,
+        }
+    }
+}
+
+impl<C: ValidConfig> AsMut<Uart<C, Duplex>> for (&mut Uart<C, RxDuplex>, &mut Uart<C, TxDuplex>) {
+    #[inline]
+    fn as_mut(&mut self) -> &mut Uart<C, Duplex> {
+        // SAFETY: Pointer casting &mut Uart<C, RxDuplex> into &mut
+        // Uart<C, Duplex> should be safe as long as RxDuplex and Duplex
+        // are both zero-sized.
+        unsafe { &mut *(self.0 as *mut _ as *mut Uart<C, Duplex>) }
```

Comment:
I think it's true that *any* struct with only one non-zero sized field has the same memory layout as the field, even for `repr(Rust)`. I remember digging into that one time. So this should be fine.